src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -2,18 +2,171 @@
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
 
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+
+    private Node<K, V>[] buckets;
+    private int size;
+
+    public MyHashMap() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    public MyHashMap(int capacity) {
+        this.buckets = new Node[capacity];
+    }
+
     @Override
     public void put(K key, V value) {
-
+        if (key == null) {
+            putNullKey(value);
+            return;
+        }
+        int index = getIndex(key);
+        Node<K, V> entry = buckets[index];
+        Node<K, V> prev = findNodeByKey(entry, key);
+        if (prev != null) {
+            prev.setValue(value);
+        } else {
+            Node<K, V> newEntry = new Node<>(key, value);
+            addToBucket(newEntry, index);
+            size++;
+        }
+        resizeIfNeeded();
     }
 
     @Override
     public V getValue(K key) {
-        return null;
+        if (key == null) {
+            return getValueForNullKey();
+        }
+        int index = getIndex(key);
+        Node<K, V> node = findNodeByKey(buckets[index], key);
+        return (node != null) ? node.getValue() : null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private void putNullKey(V value) {
+        int index = getNullKeyIndex();
+        Node<K, V> entry = buckets[index];
+        Node<K, V> prev = findNullKeyNode(entry);
+        if (prev != null) {
+            prev.setValue(value);
+        } else {
+            Node<K, V> newEntry = new Node<>(null, value);
+            addToBucket(newEntry, index);
+            size++;
+        }
+        resizeIfNeeded();
+    }
+
+    private Node<K, V> findNodeByKey(Node<K, V> entry, K key) {
+        while (entry != null) {
+            if (entry.getKey() == null || entry.getKey().equals(key)) {
+                return entry;
+            }
+            entry = entry.getNext();
+        }
+        return null;
+    }
+
+    private Node<K, V> findNullKeyNode(Node<K, V> entry) {
+        while (entry != null) {
+            if (entry.getKey() == null) {
+                return entry;
+            }
+            entry = entry.getNext();
+        }
+        return null;
+    }
```

Comment:
let's merge these two methods, they are almost the same. I'd recommend you use Objects.equals to compare your keys. Also let's merge put() and putNullKeys methods. Please merge all methods that are related to any action both for regular and null values